### PR TITLE
408 fix recheck

### LINKED ISSUE #408: `recheck` always fails
I have the following test:

```fsharp
    testCase "my test" <| fun () ->
        property {
            let! i = Gen.int32 (Range.linear 1 10)
            Expect.isTrue (i <= 5)
        }
        |> Property.check
```

It fails (as expected) with the following error:

```
    *** Failed! Falsifiable (after 60 tests):
    6
    Expecto.AssertException: Should be true. Actual value was false but had expected it to be true.
       at Expecto.Expect.isTrue@397.Invoke(String msg)
       at Expecto.Expect.isTrue(Boolean actual, String message)
       at Hedgehog.Tests.TestDsl.Expect.isTrue(Boolean value) in C:\Code\fsharp-hedgehog\tests\Hedgehog.Tests\TestDsl.fs:line 38
       at Hedgehog.Tests.PropertyTests.propertyTests@103-30.Invoke(Int32 _arg8) in C:\Code\fsharp-hedgehog\tests\Hedgehog.Tests\PropertyTests.fs:line 103
       at Hedgehog.Property.kTry@88.Invoke(a a) in C:\Code\fsharp-hedgehog\src\Hedgehog\Property.fs:line 89
    This failure can be reproduced by running:
    > Property.recheck "59_15683818216235017248_6890634259096173777_" <property>
```

I plug the suggested `recheck` into the test:

```fsharp
    testCase "my test" <| fun () ->
        property {
            let! i = Gen.int32 (Range.linear 1 10)
            Expect.isTrue (i <= 5)
        }
        |> Property.recheck "59_15683818216235017248_6890634259096173777_"
```

It fails with

```

    *** Failed! Falsifiable (after 1 test):
    6
    Expecto.AssertException: Should be true. Actual value was false but had expected it to be true.
       at Expecto.Expect.isTrue@397.Invoke(String msg)
       at Expecto.Expect.isTrue(Boolean actual, String message)
       at Hedgehog.Tests.TestDsl.Expect.isTrue(Boolean value) in C:\Code\fsharp-hedgehog\tests\Hedgehog.Tests\TestDsl.fs:line 38
       at Hedgehog.Tests.PropertyTests.propertyTests@103-30.Invoke(Int32 _arg8) in C:\Code\fsharp-hedgehog\tests\Hedgehog.Tests\PropertyTests.fs:line 103
       at Hedgehog.Property.kTry@88.Invoke(a a) in C:\Code\fsharp-hedgehog\src\Hedgehog\Property.fs:line 89
```

Great! Let's now make the test pass (trivially):

```fsharp
    testCase "my test" <| fun () ->
        property {
            let! i = Gen.int32 (Range.linear 1 10)
            Expect.isTrue true
        }
        |> Property.recheck "59_15683818216235017248_6890634259096173777_"
```

But it doesn't pass!

```
    *** Failed! Falsifiable (after 1 test):
    6
```

Is this intended behavior?

### PR DESCRIPTION
**Pull request by TysonMN**
Fixes #408

### CODE REVIEW

**Comment by TysonMN** [comment]
@AlexeyRaga, can you review?  In particular, can you test this with the code that lead you to write https://github.com/hedgehogqa/fsharp-hedgehog/issues/408#issuecomment-1191296726?

File: tests/Hedgehog.Tests/PropertyTests.fs [hunk 1]
```diff
@@ -97,6 +97,36 @@ let propertyTests = testList "Property tests" [
                 count =! 1
                 //render.Contains "actual: 1" =! true // comment out for now since it causes the Fable test to fail
 
+
+    testCase "recheck passes after code is fixed" <| fun () ->
+        let expectedShrunkenInput = 1
+        let mutable target = 0
+        let prop =
+            property {
+                let! i = Range.constant 0 1_000_000 |> Gen.int32
```
Comment (AlexeyRaga) [inline tests/Hedgehog.Tests/PropertyTests.fs:106]: Since the range starts with `0`, can't it generate `i = 0` which would be `= target`?

**Comment by AlexeyRaga** [comment]
@TysonMN looks good to me, thanks for fixing it!

**Comment by TysonMN** [comment]
@AlexeyRaga, you keep deleting your comments ;)

I had a good response typed up for your most recent one, so here it is.

> Since the range starts with `0`, can't it generate `i = 0` which would be `= target`?

Yes, and that happens.  You can put a breakpoint there to confirm verify.

An integral generator always generates the "smallest" (in terms of shrinking) value first, which in this case is `0`.  But the property is checked 100 times (by default, which is what is being used here), so the second value generated is unlikely to be `0` again (there is only `1` chance in `1,000,000` of that).  This test will fail if `0` is generated every time.  The probably of that is `(1 / 1,000,000)^99`, so this test will never fail.

**Comment by AlexeyRaga** [comment]
@TysonMN yes, I understand that. I initially just didn't understand the fact that you _wanted_ that test to fail when you run it first time :) But then I read further, understood what is going on and quickly deleted my comment hoping that you won't see my initial stupidity :)

**Comment by AlexeyRaga** [comment]
Could you please coin a release once it is merged?

**Comment by TysonMN** [comment]
Definitely.  I talked to @dharmaturtle, and he will review this tomorrow.

File: src/Hedgehog/Property.fs [hunk 2]
```diff
@@ -163,11 +163,15 @@ module Property =
...
-                }
+                let journal, outcome = root.Value
+                match outcome with
+                | Failure ->
+                    { Shrinks = 0<shrinks>
+                      Journal = journal
+                      RecheckInfo = None }
+                    |> Failed
+                | Success _ -> OK
+                | Discard -> failwith "Unexpected 'Discard' result when rechecking"
```
Comment (dharmaturtle) [inline src/Hedgehog/Property.fs:174]:
```suggestion
                | Discard -> failwith "Unexpected 'Discard' result when rechecking. This should never happen."
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+## Version 0.12.3 (2022-07-21)
+
+- Fix bug in `Property.recheck` where the result is always failed.
```
Comment (dharmaturtle) [inline CHANGELOG.md:3]:
```suggestion
- Fix bug in `Property.recheck` where the result is always `Failed`. ([#408][408], [@TysonMN][TysonMN])
```

**Review by dharmaturtle — APPROVED** [review]
Passes in my use case as well. A few very minor suggestions, but otherwise LGTM!

As side commentary, I'm not fond of making the test dependent on the shrinking behavior, but I don't feel strongly about it. One possible solution is to steal from my example in #408:

```fsharp
        let mutable target = false
        property {
            let! _ = // some random unused gen
            Expect.isTrue target
        }
```

But again, I don't feel strongly about this.

**Comment by TysonMN** [comment]
> I'm not fond of making the test dependent on the shrinking behavior, but I don't feel strongly about it. One possible solution is to steal from my example in #408:

Yes, must better.  I changed to your suggestion.

**Pull request merged**